packages/integrations/components/coral/usePicoObserver.js

```diff
@@ -0,0 +1,138 @@
+  const dispatchLogoutRequest = useCallback(
+    () => dispatch(actionReceiveCoralLogoutRequest()),
+    [dispatch]
+  );
+  // Define a function to dispatch a plan upgrade requirement message.
+  const dispatchReceivePlanUpgrade = useCallback(
+    () => dispatch(actionReceivePicoPlanUpgrade()),
+    [dispatch]
+  );
+  // Define an variable used to signify whether a user is already authenticated
+  // with Pico and is required to upgrade their subscription prior to being
+  // sent through the Coral SSO workflow.
+  const showUpgradeModal = useSelector(showUpgradeModalSelector);
+
+  // Mount the effect handler.
+  useEffect(() => {
+    const observerHandler = () => {
+      const signalNode = document.getElementById('PicoSignal-container');
+      // Only mount the observer if the PicoSignal button exists in the DOM.
+      if (signalNode) {
+        // Define the observer and debounce to prevent too many updates firing at once.
+        const observer = new MutationObserver(debounce((mutations) => {
+          mutations.forEach((mutation) => {
+            const { type, attributeName } = mutation;
+            // Execute the webhook on the `data-pico-status` attribute.
+            if ('attributes' === type && 'data-pico-status' === attributeName) {
+              // Create a reusable store for attribute values.
+              let attributes = {
+                status: signalNode.getAttribute('data-pico-status'),
+                email: signalNode.getAttribute('data-pico-email'),
+                tier: signalNode.getAttribute('data-pico-tier'),
+              };
+
+              // Once the `data-pico-status` and `data-pico-email` attributes are
+              // populated, send a request to Pico to verify the user and retrieve
+              // a JWT for logging the user into Coral via SSO.
+              if (
+                0 < attributes.email.length &&
+                0 < tiers.length &&
+                (
+                  'registered' === attributes.status ||
+                  'paying' === attributes.status
+                )
+              ) {
+                // Ensure the user's current tier matches the levels available
+                // for Coral SSO.
+                if (! tiers.includes(attributes.tier) && ! showUpgradeModal) {
+                  dispatchRequireUpgradeMessage();
+                  observer.disconnect();
+                } else {
+                  // Dispatch an action that signifies the user has upgraded
+                  // their subscription to a level sufficient to enable Coral SSO.
+                  if (showUpgradeModal) {
+                    dispatchReceivePlanUpgrade();
+                  }
+
+                  if (! requestSent) {
+                    // Set a delay in order for the global Pico object to update with
+                    // the user's ID prior to dispatching the verification request.
+                    setTimeout(() => {
```

Comment:
Oh, fun race conditions! Since we don't care about IE11, I wonder if this would be a good use case for using a [JS Proxy object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)?

Reply:
Interesting thought. I'll give that implementation some experimentation time in DEF-235